### PR TITLE
feat: add `Finish` stage

### LINKED ISSUE #1262: Add Finish stage
### Describe the feature

The `Finish` stage marks the highest fully synced block which should be communicated to the network and the RPC. The highest fully synced block should be checked for on start up of the node.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by onbjerg**
Adds the `Finish` stage and adds logic to use the correct block in RPC and P2P messages

Closes #1262

### CODE REVIEW

File: crates/stages/src/stages/finish.rs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    fn id(&self) -> StageId {
+        FINISH
+    }
+
+    async fn execute(
+        &mut self,
+        _: &mut Transaction<'_, DB>,
+        input: ExecInput,
+    ) -> Result<ExecOutput, StageError> {
+        Ok(ExecOutput { done: true, stage_progress: input.previous_stage_progress() })
```
Comment (gakonst) [inline crates/stages/src/stages/finish.rs:27]: Does this need to call update_head somehow? Or trigger something downstream that updates the head for the p2p status msg

Comment (onbjerg) [reply crates/stages/src/stages/finish.rs:27]: Yes, that's why it's a draft

File: crates/stages/src/sets.rs [hunk 2]
```diff
@@ -51,27 +54,42 @@ use std::sync::Arc;
...
     H: HeaderDownloader + 'static,
     B: BodyDownloader + 'static,
+    S: StatusUpdater + 'static,
 {
     fn builder(self) -> StageSetBuilder<DB> {
-        self.online.builder().add_set(OfflineStages)
+        self.online
+            .builder()
+            .add_set(OfflineStages)
+            .add_stage(FinishStage::new(self.status_updater))
```
Comment (rkrasiuk) [inline crates/stages/src/sets.rs:92]: in the `Finish` stage you rely on total difficulty values, but TD stage is not a part of default stages. we should add it

**Review by rkrasiuk — CHANGES_REQUESTED** [review]
look good overall, leaving a couple of comments/questions and one change request for including TD in default stages

File: crates/stages/src/stages/finish.rs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+impl<S> FinishStage<S>
+where
+    S: StatusUpdater,
+{
+    /// Create a new [FinishStage] with the given [StatusUpdater].
+    pub fn new(updater: S) -> Self {
+        Self { updater }
+    }
+
+    /// Construct a [Head] POD from `block_number`.
```
[the remark this answers is not in the document]
Comment (onbjerg) [reply crates/stages/src/stages/finish.rs:34]: plain old data

**Review by mattsse — APPROVED** [review]
as discussed for announcing the synced tip this makes perfectly sense, mainly required by the p2p network for updating the forkfilter etc.

File: crates/storage/provider/src/providers/mod.rs [hunk 4]
```diff
@@ -76,12 +76,13 @@ impl<DB: Database> BlockHashProvider for ShareableDatabase<DB> {
...
-            last_finalized: None,
-            safe_finalized: None,
-        })
+        let best_number = self
+            .db
+            .view(|tx| tx.get::<tables::SyncStage>("Finish".as_bytes().to_vec()))?
+            .map_err(Into::<reth_interfaces::db::Error>::into)?
+            .unwrap_or_default();
+        let best_hash = self.block_hash(U256::from(best_number))?.unwrap_or_default();
+        Ok(ChainInfo { best_hash, best_number, last_finalized: None, safe_finalized: None })
```
[the remark this answers is not in the document]
Comment (gakonst) [reply crates/storage/provider/src/providers/mod.rs:85]: cc @mattsse ^

**Pull request merged**